File: web/static/js/game_channel.js
import $ from "jquery";
import socket from "./socket";

let gameId = window.gameId;

function init(gameId) {
  let channel = socket.channel(`game:${gameId}`, {});
  let gameBoard = $("#gameBoard");

  function pauseGame(e) {
    channel.push("pause", {})
  }

  function startGame(e) {
    channel.push("start", {})
  }

  function stopGame(e) {
    channel.push("stop", {})
  }

  function nextTurn(e) {
    channel.push("next", {})
  }

  function prevTurn(e) {
    channel.push("prev", {})
  }

  function replay(e) {
    channel.push("replay", {})
  }

  function handleTick({html}) {
    gameBoard.html(html);
  }

  function keydownHandler(event) {
    switch (event.key) {
      case "r":
        startGame(event);
        break;
      case "R":
        replay(event);
        break;
      case " ":
        pauseGame(event);
        break;
      case "q":
        stopGame(event);
        break;
      case "j":
        nextTurn(event);
        break;
      case "ArrowRight":
        nextTurn(event);
        break;
      case "k":
        prevTurn(event);
        break;
      case "ArrowLeft":
        prevTurn(event);
        break;
    }
  }

  $(document).on("click", '[data-js="game.start"]', startGame);
  $(document).on("click", '[data-js="game.pause"]', pauseGame);
  $(document).on("click", '[data-js="game.stop"]', stopGame);
  $(document).on("click", '[data-js="game.next"]', nextTurn);
  $(document).on("click", '[data-js="game.prev"]', prevTurn);
  $(document).on("keydown", "body", keydownHandler);

  // replace the board on each tick
  channel.on("tick", handleTick);

  channel.join().
    receive("ok", resp => { console.log("Joined channel", resp) }).
    receive("error", resp => { console.log("Unable to join", resp) });
}

if (typeof gameId !== 'undefined') {
  init(gameId);
}

File: web/channels/game_channel.ex
defmodule BattleSnake.GameChannel do
  alias BattleSnake.GameServer
  alias BattleSnake.GameServer.State.Event

  use BattleSnake.Web, :channel

  import Phoenix.View, only: [render_to_string: 3]

  @spec join(binary, %{}, Phoenix.Socket.t) ::
  {:ok, Phoenix.Socket.t} |
  {:error, any}
  def join("game:" <> game_id, payload, socket) do
    if authorized?(payload) do
      socket = assign(socket, :game_id, game_id)

      with :ok <- GameServer.PubSub.subscribe(game_id),
           {:ok, pid} <- GameServer.Registry.lookup_or_create(game_id) do
        socket = assign(socket, :game_server_pid, pid)
        {:ok, socket}
      end

    else
      {:error, %{reason: "unauthorized"}}
    end
  end

  @spec handle_info(:after_join, Phoenix.Socket.t) ::
  {:noreply, Phoenix.Socket.t}
  def handle_info(:after_join,
    %{assigns: %{game_server_pid: game_server_pid}}
    = socket) do
    state = GameServer.get_status(game_server_pid)
    push(socket, "state_change", %{data: state})
    {:noreply, socket}
  end

  def handle_info(:after_join, socket) do
    {:noreply, socket}
  end

  @doc """
  Handle notifications from the GameSever. This is used to render the state of
  the game as an svg and broadcast it to all channel subscribers.
  """
  def handle_info(%Event{name: name, data: state}, socket) do
    case name do
      _ ->
        html = render_board(state)
        broadcast(socket, "tick", %{html: html})
    end
    {:noreply, socket}
  end

  def handle_in("replay", _, socket) do
    game_server_pid = socket.assigns.game_server_pid
    :ok = GameServer.replay(game_server_pid)
    {:reply, :ok, socket}
  end

  def handle_in("start", _, socket) do
    game_server_pid = socket.assigns.game_server_pid
    :ok = GameServer.resume(game_server_pid)
    {:reply, :ok, socket}
  end

  def handle_in("pause", _, socket) do
    :ok = socket.assigns.game_server_pid
    |> GameServer.pause()
    {:reply, :ok, socket}
  end

  def handle_in("next", _, socket) do
    socket.assigns.game_server_pid
    |> GameServer.next()
    {:reply, :ok, socket}
  end

  def handle_in("stop", _, socket) do
    game_server_pid = socket.assigns.game_server_pid
    :ok = GenServer.stop(game_server_pid, :normal)
    {:reply, :ok, socket}
  end

  def handle_in("prev", _, socket) do
    socket.assigns.game_server_pid
    |> GameServer.prev()
    {:reply, :ok, socket}
  end

  defp authorized?(_payload) do
    true
  end

  defp render_board(state),
    do: render_to_string(BattleSnake.PlayView, "board.html", state: state)
end
